lib/Dialect/FIRRTL/Transforms/ExtractInstances.cpp

```diff
@@ -0,0 +1,964 @@
+    for (auto *instRecord : *module) {
+      auto *target = instRecord->getTarget();
+      if (dutModules.insert(target->getModule()).second)
+        worklist.push_back(target);
+    }
+  }
+
+  // If clock gate extraction is requested, find instances of extmodules with
+  // the corresponding `defname` and mark them as to be extracted.
+  // TODO: This defname really shouldn't be hardcoded here. Make this at least
+  // somewhat configurable.
+  if (!clkgateFileName.empty()) {
+    auto clkgateDefNameAttr = StringAttr::get(&getContext(), "EICG_wrapper");
+    for (auto module : circuit.getOps<FExtModuleOp>()) {
+      if (module.defnameAttr() != clkgateDefNameAttr)
+        continue;
+      LLVM_DEBUG(llvm::dbgs()
+                 << "Clock gate `" << module.moduleName() << "`\n");
+      if (!dutModules.contains(module)) {
+        LLVM_DEBUG(llvm::dbgs() << "- Ignored (outside DUT)\n");
+        continue;
+      }
+
+      ExtractionInfo info;
+      info.traceFilename = clkgateFileName;
+      info.prefix = "clock_gate"; // TODO: Don't hardcode this
+      info.wrapperModule = clkgateWrapperModule;
+      info.stopAtDUT = !info.wrapperModule.empty();
+      for (auto *instRecord :
+           instanceGraph->lookup(cast<hw::HWModuleLike>(*module))->uses()) {
+        if (auto inst = dyn_cast<InstanceOp>(*instRecord->getInstance())) {
+          LLVM_DEBUG(llvm::dbgs()
+                     << "- Marking `"
+                     << inst->getParentOfType<FModuleLike>().moduleName() << "."
+                     << inst.name() << "`\n");
+          extractionWorklist.push_back({inst, info});
+        }
+      }
+    }
+  }
+
+  // TODO: Enable the following once memory-to-module lowering exists.
+  // If memory extraction is requested, find instances of memory modules and
+  // mark them as to be extracted.
+  // if (!memoryFileName.empty()) {
+  //   for (auto module : circuit.getOps<FModuleMemOp>()) {
+  //     LLVM_DEBUG(llvm::dbgs() << "Memory `" << module.moduleName() << "`\n");
+  //     if (!dutModules.contains(module)) {
+  //       LLVM_DEBUG(llvm::dbgs() << "- Ignored (outside DUT)\n");
+  //       continue;
+  //     }
+
+  //     ExtractionInfo info;
+  //     info.traceFilename = memoryFileName;
+  //     info.prefix = "mem_wiring"; // TODO: Don't hardcode this
+  //     info.wrapperModule = memoryWrapperModule;
+  //     info.stopAtDUT = !info.wrapperModule.empty();
+  //     for (auto *instRecord :
+  //          instanceGraph->lookup(cast<hw::HWModuleLike>(*module))->uses()) {
+  //       if (auto inst = dyn_cast<InstanceOp>(*instRecord->getInstance())) {
+  //         LLVM_DEBUG(llvm::dbgs()
+  //                    << "- Marking `"
+  //                    << inst->getParentOfType<FModuleLike>().moduleName() <<
+  //                    "."
+  //                    << inst.name() << "`\n");
+  //         extractionWorklist.push_back({inst, info});
+  //       }
+  //     }
+  //   }
+  // }
```

Comment:
This could be factored out into a branch before merging.

Reply:
Good idea, will do! :+1: